### PR TITLE
convert Repliear to use the row versioning strategy

### PR DESCRIPTION
**Pull request by tantaman**
Based on `todo-row-versioning`. This won't diff well since it re-structured the app and switched to Vite.

Most relevant files:

### Server
- [schema.ts](https://github.com/tantaman/repliear-row-versioning/blob/main/server/src/schema.ts) - DB Schema
- [pull.ts](https://github.com/tantaman/repliear-row-versioning/blob/main/server/src/pull/pull.ts) - Pull implementation.
- [cvr.ts](https://github.com/tantaman/repliear-row-versioning/blob/main/server/src/pull/cvr.ts) - Functions to pull data, diff against and populate a CVR

Note that past CVRs contribute to the current CVR. I.e., all data the client currently has is represented by all data we've ever sent. If any CVR is dropped we'll just re-fetch the rows that the CVR used to have.

### Client
- [app.tsx#sync](https://github.com/tantaman/repliear-row-versioning/blob/eb136c96ba34808cb6c4b5569773c41ac72d91c0/client/src/app.tsx#L53-L93) - pull chunks of issues/comments/descriptions until we're done
- [mutators.ts](https://github.com/tantaman/repliear-row-versioning/blob/main/client/src/model/mutators.ts) - mutators using Rails
- [issue-detail.ts](https://github.com/tantaman/repliear-row-versioning/blob/main/client/src/issue/issue-detail.tsx) - loading data via Rails

Note that I'm currently only allowing one tab to pull at once. This is explained in `pull.ts` in the comment above `dropCVREntries`. This is something I could fix if we want to keep the behavior of allowing tabs to sync the same data at the same time.

### CODE REVIEW

File: server/src/pull/pull.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        clientVersion: 0,
+        order: 0,
+      };
+
+      // Drop any cvr entries greater than the order we received.
+      // Getting an old order from the client means that the client is possibly missing the data
+      // from future orders.
+      //
+      // The other possbility is that multiple tabs are syncing (why are tabs not coordinating who does sync?)
+      // Given that, dropping entries can cause to tabs could starve one another!! I've updated the frontend
```
Comment (aboodman) [inline server/src/pull/pull.ts:59]: I do not think they can starve each other actually. The snapshots are shared between clients, just slowly (currently ~2s I think). So if you have tab A and B, and A receives order 1, and B 2. Then A pulls again and the server sees that 2 is ahead, it will go ahead and delete and respond to A. But then A will share this snapshot with B and get the new cookie. When B next pulls, he will use the cookie that A received and all is well.

Comment (aboodman) [reply server/src/pull/pull.ts:59]: See also: https://www.notion.so/replicache/DD-3-1-e42489fc2e6b4340a01c7fa0de353a30

File: server/src/pull/next-page.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    {rows: issueDeleteRows},
+    {rows: descriptionDeleteRows},
+    {rows: commentDeleteRows},
+  ] = await Promise.all([
+    findDeletions(executor, 'issue', clientGroupID, order, remaining),
+    findDeletions(executor, 'description', clientGroupID, order, remaining),
+    findDeletions(executor, 'comment', clientGroupID, order, remaining),
+  ]);
+
+  // Record deletes against the _current_ CVR
```
Comment (aboodman) [inline server/src/pull/next-page.ts:164]: This sounds important but I'm not getting the distinction. The `findUnsentItems` function finds rows that are new. Doesn't this do the same for deletions that are new?

File: server/src/pull/pull.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+      // so pulls can only be issued from one tab (see `useExclusiveEffect`). The other solution here is to not
```
Comment (aboodman) [inline server/src/pull/pull.ts:60]: We looked at web locks extensively and decided not to rely on them. The semantics around tab freezing, bfcache, and similar are super _super_ weird, and we have low confidence they actually work right and across browsers.

I think we should continue to allow multiple tabs to sync because it's robust, but we should prevent _background_ tabs from syncing.

I don't think the starvation thing you mention will happen.

**Review by aboodman — COMMENTED** [review]
Still thinking about the big picture here, will review more tomorrow.

I like the representation of the CVR in the database, very clever. I wonder how much you thought about its performance as it grows more and more spread out. Does it ever need to be compacted?

Some other comments about the design of pull inside.

**Comment by tantaman** [comment]
> I wonder how much you thought about its performance as it grows more and more spread out. Does it ever need to be compacted?

I believe that `SELECT rowid, rowversion WHERE NOT IN (...)` will slow down as the CVR gets larger. I don't know by how much or where the limit is so that is something that:

a. I can dig into more or
b. we can ignore for now since we're going to replace this "sync the world" with UI driven sync which may end up changing the type of query we issue


side note: I was very surprised that `WHERE NOT IN (...)` performed better than `WHERE NOT EXISTS`.

File: server/src/schema.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  )`);
+
+  await executor(/*sql*/ `CREATE TABLE "cvr_entry" (
+    "client_group_id" VARCHAR(36) NOT NULL,
+    "order" INTEGER NOT NULL,
+    "tbl" INTEGER NOT NULL,
+    "row_id" VARCHAR(36) NOT NULL,
+    "row_version" INTEGER NOT NULL,
+    -- unique by client_group_id, tbl, row_id
+    -- 1. A missing row version is semantically the same as a behind row version
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:87]: This might also be related to your compaction question.

I'm only keeping a single entry per `row_id`.

If on the first pull we have cvr entries that would look like:


|client_group|order|row_id|row_version|
|------------|-----|-------|------------|
|1 | 1 | 1 | 1 |
|1 | 1| 2 | 1 |

And on a second pull:

|client_group|order|row_id|row_version|
|------------|-----|-------|------------|
|1 | 2 | 1 | 2 |

Then the table, after both pulls are written, would look like:

|client_group|order|row_id|row_version|
|------------|-----|-------|------------|
| 1 | 1 | 2 | 1 |
|1 | 2 | 1 | 2 |

Notice that we are missing the `|1 | 1 | 1 | 1 |` entry. This is because the second pull caused it to be evicted from the first pull's CVR. This is fine since the `row_version` is ahead of what the first CVR had and a missing entry is treated the same as a behind entry.

File: server/src/pull/pull.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        clientGroupID,
+        clientVersion: 0,
+        order: 0,
+      };
+
+      // Drop any cvr entries greater than the order we received.
+      // Getting an old order from the client means that the client is possibly missing the data
+      // from future orders.
+      //
+      // The other possbility is that multiple tabs are syncing (why are tabs not coordinating who does sync?)
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/pull/pull.ts:58]: I thought you'd have to do leader election for my work on cr-sqlite in the browser as well but it seems like WebLocks just solve this? Have each tab try to acquire the lock as its first operation, once the lock is acquired that tab starts to sync. When tabs die, this'll free the lock for other tabs to take over sync.

The only question is if a tab can be put to sleep while holding a lock. I haven't seen this in practice and that sounds like bad design if browsers could sleep tabs that hold shared resources.

**Comment by aboodman** [comment]
On Thu, Nov 16, 2023 at 4:54 AM Matt Wonlaw ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In server/src/pull/pull.ts
> <https://github.com/rocicorp/repliear/pull/158#discussion_r1395817001>:
>
> > +        clientGroupID,
> +      );
> +      const prevCVR = pull.cookie
> +        ? await getCVR(executor, pull.cookie.clientGroupID, pull.cookie.order)
> +        : undefined;
> +      const baseCVR: CVR = prevCVR ?? {
> +        clientGroupID,
> +        clientVersion: 0,
> +        order: 0,
> +      };
> +
> +      // Drop any cvr entries greater than the order we received.
> +      // Getting an old order from the client means that the client is possibly missing the data
> +      // from future orders.
> +      //
> +      // The other possbility is that multiple tabs are syncing (why are tabs not coordinating who does sync?)
>
> I thought you'd have to do leader election for my work on cr-sqlite in the
> browser as well but it seems like WebLocks just solve this? Have each tab
> try to acquire the lock as its first operation, once the lock is acquired
> that tab starts to sync. When tabs die, this'll free the lock for other
> tabs to take over sync.
>

We spent a ton of time trying to decide if we could trust web locks and
couldn't convince ourselves.

Here are some starting points:

https://x.com/aboodman/status/1549590251558936577?s=20 ***@***.*** in that
thread is ex-mozilla)
https://github.com/w3c/web-locks/issues/81

I would love to be convinced, but there are so many weird edge cases in
browsers around bfcache, frozen states – especially in mobile where it's
even harder to test, that it's very hard to feel confident – particularly
since the spec doesn't even say what should happen yet.

Everytime we get anywhere near here I just give up and decide it's easier
to engineer something that works right without locks than try to understand
if these things work across all the browsers and platforms we want to
support.

Luckily, I don't think web locks are necessary here -- I believe your
big-picture design does work without them:
https://github.com/rocicorp/repliear/pull/158#discussion_r1395524886

Am I missing something?

File: server/src/pull/next-page.ts (hunk 2 above)
Comment (tantaman) [reply server/src/pull/next-page.ts:164]: This was a comment from a prior iteration where I was thinking through ideas to be able to remove tombstones.

Will remove the comment.

File: client/index.html [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <meta charset="UTF-8" />
+  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+  <title>Replicache Todo</title>
```
Comment (grgbkr) [inline client/index.html:7]:
```suggestion
  <title>Repliear</title>
```

File: server/src/pull/pull.ts (hunk 1 above)
Comment (tantaman) [reply server/src/pull/pull.ts:59]: > The snapshots are shared between clients, just slowly (currently ~2s I think)

So maybe not starving one another but sounds like there'll be a lot of thrown away syncs while we wait the 2 seconds for the snapshots to get shared.

While snapshot isn't shared:

```
Starting state: server order 2, A order 1, B order 2

A pulls order 1
Server: delete 2, send order 3 to A
B pulls order 2
Server: delete 3, send order 4 to B
A pulls order 3
Server: delete 4, send order 5 to A
```

3 is based on order 1.
4 is based on order 1.
5 is based on order 1.

Nobody made progress in that cycle.

Will take my lock out and test though.

**Comment by tantaman** [comment]
> Am I missing something?

No, I don't think so. You've definitely done more research on weblocks than myself.

File: server/src/pull/pull.ts (hunk 3 above)
Comment (tantaman) [reply server/src/pull/pull.ts:60]: > but we should prevent background tabs from syncing.

Just via something like `document.hidden`?

Comment (aboodman) [reply server/src/pull/pull.ts:60]: Yes, that's what I'm thinking.

File: server/src/pull/pull.ts (hunk 1 above)
Comment (aboodman) [reply server/src/pull/pull.ts:59]: Yeah, I agree this is a problem now. A design for the `cvr_entry` table that would be "compatible" with the current design of Replicache would be one where older rows are retained instead of mutated in place. Something like:

```
cvr_entry
=======
client_group_id: int, fk(client_group.id)
tbl: enum
entity_version: int
valid_from_client_view_version: int
valid_to_client_view_version: int nullable # null means "still valid"
```

When an entry in the client view is modified, we:
- set the previous entry's `valid_to_client_view_version`
- create a new version with the appropriate `valid_from_client_view_version`

Alternately, perhaps @grgbkr can remind me one more time why this delay is here. I remember we were trying to ensure that the network sync usually won, but now I can't remember why! Perhaps it would be OK to tweak this so that local sync usually wins. In this case your current design would work.

File: server/src/pull/cvr.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  order: number,
+  updates: SearchResult[],
+) {
+  if (updates.length === 0) {
+    return;
+  }
+
+  const placeholders = [];
+  const values = [];
+  const stride = 5;
```
Comment (aboodman) [inline server/src/pull/cvr.ts:154]: Why build up these two huge arrays. Can you instead allocate a single array of length five for the values and use it over and over? There's no need for the placeholder array in the first place -- you can just hardcode the placeholders in the SQL string.

File: server/src/pull/pull.ts (hunk 1 above)
Comment (aboodman) [reply server/src/pull/pull.ts:59]: > but sounds like there'll be a lot of thrown away syncs while we wait the 2 seconds for the snapshots to get shared.
...
> 3 is based on order 1.
> 4 is based on order 1.
> 5 is based on order 1.
> Nobody made progress in that cycle.

Wait, I don't agree nobody makes progress.

It's true that in your example the response with order `4` will be based on order `1`. But it's not true that the client requesting it makes no progress. He gets synced to head. It's just that the delta is bigger than necessary: he had up through order `2`, but the patch included those changes even though he didn't need them. This continues with any subsequent pulls until the tabs share their state.

So I think your design is robust without the lock, just slightly inefficient.

This should of course be tested to confirm.

File: server/src/schema.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    await createSchemaVersion1(executor);
+  }
+}
+
+export async function createSchemaVersion1(executor: Executor) {
+  await executor(
+    'CREATE TABLE replicache_meta (key text PRIMARY KEY, value json)',
+  );
+  await executor(
+    "insert into replicache_meta (key, value) values ('schemaVersion', '1')",
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:20]: resolved in 42726a853bd91312ba394fd918e6465df16a0147

File: server/src/schema.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  await executor(/*sql*/ `CREATE TABLE "issue" (
+    "id" VARCHAR(36) PRIMARY KEY NOT NULL,
+    "title" text NOT NULL,
+    "priority" priority,
+    "status" status,
+    "modified" BIGINT NOT NULL,
+    "created" BIGINT NOT NULL,
+    "creator" VARCHAR(36) NOT NULL,
+    "kanbanorder" VARCHAR(36),
+    "rowversion" INTEGER NOT NULL
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:55]: resolved in 5533cd096b1cea8255764cf9137f73314bd5eb5a

File: server/src/schema.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  )`);
+
+  await executor(/*sql*/ `CREATE TABLE "description" (
+    "id" VARCHAR(36) PRIMARY KEY NOT NULL REFERENCES issue("id") ON DELETE CASCADE,
+    "body" text NOT NULL,
+    "rowversion" INTEGER NOT NULL
+  )`);
+
+  await executor(/*sql*/ `CREATE TABLE "cvr" (
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:73]: resolved in 4f9e1a7bd240b811229e181f5d7477e9b8b4edd5

File: server/src/schema.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    "client_group_id" VARCHAR(36) NOT NULL,
+    "order" INTEGER NOT NULL,
+    "client_version" INTEGER NOT NULL,
+    PRIMARY KEY ("client_group_id", "order")
+  )`);
+
+  await executor(/*sql*/ `CREATE TABLE "cvr_entry" (
+    "client_group_id" VARCHAR(36) NOT NULL,
+    "order" INTEGER NOT NULL,
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:82]: resolved in 312e9f49c6fec78d2e17871d8faf229630118df3

File: server/src/schema.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  await executor(/*sql*/ `CREATE TABLE "description" (
+    "id" VARCHAR(36) PRIMARY KEY NOT NULL REFERENCES issue("id") ON DELETE CASCADE,
+    "body" text NOT NULL,
+    "rowversion" INTEGER NOT NULL
+  )`);
+
+  await executor(/*sql*/ `CREATE TABLE "cvr" (
+    "client_group_id" VARCHAR(36) NOT NULL,
+    "order" INTEGER NOT NULL,
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:75]: resolved in 312e9f49c6fec78d2e17871d8faf229630118df3

File: server/src/schema.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    "client_version" INTEGER NOT NULL,
+    PRIMARY KEY ("client_group_id", "order")
+  )`);
+
+  await executor(/*sql*/ `CREATE TABLE "cvr_entry" (
+    "client_group_id" VARCHAR(36) NOT NULL,
+    "order" INTEGER NOT NULL,
+    "tbl" INTEGER NOT NULL,
+    "row_id" VARCHAR(36) NOT NULL,
+    "row_version" INTEGER NOT NULL,
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts:85]: why not `row_version` to pair well with `row_id`?

That or rename both to `entity_id` and `entity_version`

[the remark this answers is not in the document]
Comment (aboodman) [reply server/src/schema.ts:85]: Yeah I think renaming both to `entity` make sense. This is pretty minor but I was thinking that if we would say `issue_id` and `issue_version` in the normal concrete case, then the abstract names should be `entity_id` and `entity_version`. (and perhaps we should say `entity` instead of `tbl` as well).

[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/schema.ts]: resolved in 9213d6951f9a14a14d423150fda71b0aa1831022

File: server/src/pull/cvr.ts (hunk 7 above)
Comment (tantaman) [reply server/src/pull/cvr.ts:155]: The intent was to reduce the number of calls to the DB.

E.g., if we're recording a batch of 3,000 things then it'll just issue 1 query with all 3k values.

Not building up these arrays would mean sending 3,000 separate insert queries.

E.g., changing from this:

```ts
for (let i = 0; i < updates.length; i++) {
    placeholders.push(
      `($${i * stride + 1}, $${i * stride + 2}, $${i * stride + 3}, $${
        i * stride + 4
      }, $${i * stride + 5})`,
    );
    values.push(
      clientGroupID,
      order,
      TableOrdinal[table],
      updates[i].id,
      updates[i].version,
    );
  }

  await executor(
    /*sql*/ `INSERT INTO client_view_entry (
    "client_group_id",
    "client_view_version",
    "entity",
    "entity_id",
    "entity_version"
  ) VALUES ${placeholders.join(
    ', ',
  )} ON CONFLICT ("client_group_id", "entity", "entity_id") DO UPDATE SET
    "entity_version" = excluded."entity_version",
    "client_view_version" = excluded."client_view_version"
  `,
    values,
  );
```

to this:

```ts
await Promise.all(updates.map(u => {
    await executor(
      /*sql*/ `INSERT INTO client_view_entry (
      "client_group_id",
      "client_view_version",
      "entity",
      "entity_id",
      "entity_version"
    ) VALUES (?, ?, ?, ?, ?) ON CONFLICT ("client_group_id", "entity", "entity_id") DO UPDATE SET
      "entity_version" = excluded."entity_version",
      "client_view_version" = excluded."client_view_version"
    `,
     [clientGroupID, order, table, u.id, u.version],
    );
}))
```

So it's either:
1. Create and await 3k promises (/ some chunking of this)
2. Or create a large string of 3k entries and 1 promise.

File: server/src/pull/next-page.ts (hunk 2 above)
Comment (tantaman) [reply server/src/pull/next-page.ts]: resolved ddee4f1ad4ef54fbbf868b2940a397b4069b642e

**Comment by tantaman** [comment]
https://github.com/rocicorp/repliear/pull/158/commits/9f1cb550ade7667dc30d66c4f575494150f69342 fixes things so many tabs can make progress if they're all trying to sync at once.

At this point it warrants some proper integration tests and compaction (cvrs now grow without bound).

The basic idea is that we no longer drop `client_views` nor mutate them in place. Each client view refers to its parent and we use that linked list of client views to construct the full view. If tabs branch off they get their own view that is not interfered with by other tabs.

image

File: server/src/schema.ts [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  );
+  await executor(
+    "insert into replicache_meta (key, value) values ('schemaVersion', '1')",
+  );
+
+  // cvrversion is null until first pull initializes it.
+  await executor(/*sql*/ `CREATE TABLE replicache_client_group (
+    id VARCHAR(36) PRIMARY KEY NOT NULL,
+    cvrversion INTEGER null,
+    clientversion INTEGER NOT NULL,
```
[the remark this answers is not in the document]
Comment (aboodman) [reply server/src/schema.ts:27]: The patch in Replicache contains basically three pieces of data:
1. A patch to the client view
2. A patch to lastMutationIDs
3. A cookie that encodes current state of backend db

todo-row-versioning calculates 1 and 2 using two different systems. For (1) it uses row versioning. For (2) it uses, basically, per-space versioning. Each ReplicaheClientGroup has a version and the clients have the version they updated at.

It would be more elegant to use row-versioning for calculating both patches. This would entail changing the way the version on the ReplicacheClient rows works. Instead of taking the latest value of the RCG.version, it would increment on its own independently, and then be stored as part of the CVR like other data.

The reason I didn't do this originally in todo-row-versioning is that every page load generates a new client. The set of clients only grows, and every CVR would contain the entire set of them. The in-memory diff would eventually get out of control.

But with your implementation here the diff is done inside the database using indexes. I am guessing that it could stay fast indefinitely.

The only problem with my proposal is that with it, the CVR table would grow with O(number-of-clients) rather than as it does now with O(number-of-rows-in-primary-data). But I believe that is addressable. Right now Replicache doesn't completely support deleting old client records (https://github.com/rocicorp/replicache/issues/1033). But if it did, then we could delete them, and they'd automatically get deleted from this CVR table too.

**Comment by aboodman** [comment]
> https://github.com/rocicorp/repliear/commit/9f1cb550ade7667dc30d66c4f575494150f69342 fixes things so many tabs can make progress if they're all trying to sync at once.

OK, I want to understand why it didn't work without this. Going to try and load it into my brain today.

File: server/src/pull/cvr.ts (hunk 7 above)
Comment (aboodman) [reply server/src/pull/cvr.ts:176]: I seeeeee. I totally missed this, thank you for explaining.

Is it actually faster? Did you measure it? If significantly faster, I support this complexity but if it's marginal I'd prefer the more direct `Promise.all()` impl.

Comment (aboodman) [reply server/src/pull/cvr.ts:176]: Nevermind -- https://github.com/vitaly-t/pg-promise/wiki/Performance-Boost confirms that it's likely to be much slower (and also has some helpers that may or may not be useful to you).

**Comment by aboodman** [comment]
> 9f1cb55
<https://github.com/rocicorp/repliear/commit/9f1cb550ade7667dc30d66c4f575494150f69342>
fixes
things so many tabs can make progress if they're all trying to sync at once.

Thank you for doing this, but let's please pause and figure out the design
of client_view_entry first. I've grown to really like your original design
where it's mutated in-place and doesn't grow. If we do do an immutable
design I think there's a more relational-friendly design that's possible
and doesn't require compaction.

Greg and I are meeting tomorrow midday PST to figure out our opinion and
will share it after that. Maybe we could even all meet together.

In the meantime...

>> I think with your system here we could actually implement the client
records using cvr too, which would simplify things! #158 (comment)
<https://github.com/rocicorp/repliear/pull/158#discussion_r1407002841>

> Didn't do this / waiting clarification. I think we could also merge and
do this later.

Can you look into this? It seems orthogonal to above. I think it's likely
Greg may come back with some feedback tomorrow too.

And one last thing:

The original Repliear implementation created a unique "space" for each
visitor so that they didn't interfere with each others' writes. You can see
this if you visit production:

[image: CleanShot 2023-11-30 at ***@***.***


 This was a large amount of complexity both inthe code and schema though.
Perhaps we should do something simpler like reset the data every 24h. This
could be done after merge, but registering it here as a TODO.

**Comment by aboodman** [comment]
On Thu, Nov 30, 2023 at 8:08 PM Aaron Boodman ***@***.***>
wrote:

> If we do do an immutable design I think there's a more relational-friendly
> design that's possible and doesn't require compaction.
>

To avoid leaving the begged question hanging, here's what i was thinking:

client_view_entry
-----------------------------------
client_group_id string
entity(enum)
entity_id string
entity_version integer
valid_from_client_view_version integer
valid_to_client_view_version nullable integer

To read the current client view for a cg: `select entity_id, entity_version
from client_view_entry where id = $client_group_id and
valid_from_client_view_version >= $client_group_version and
coalesce(valid_to_client_view_version, $client_group_version) <=
$client_group_version`

To update the client view for a cg (handwaving)
- Pick the new client_view version
- For each entity row either not in the client view or whose version is
higher in entity row:
- If there is an existing valid row for that entity, set its
valid_to_client_view_version to the previous client_view version
- Insert a new client_view_entry row with valid_to_client_view_version =
NULL

(this idea basically stolen from "temporal tables" but using
valid_from/to_version rather than valid_from/to_timestamp)

**Comment by tantaman** [comment]
> Greg and I are meeting tomorrow midday PST to figure out our opinion and
will share it after that. Maybe we could even all meet together.

Sure. I should be available if it's before 5pm EST.

File: server/src/pull/cvr.ts [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+  return executor(sql, params);
+}
+
+export const TableOrdinal = {
+  issue: 1,
+  description: 2,
+  comment: 3,
+} as const;
+
+export function findDeletions(
```
Comment (grgbkr) [inline server/src/pull/cvr.ts:93]: I'm not completely understanding how deletions work.

Does this correctly handle an entry being put, then deleted, then later being put again?

**Review by grgbkr — COMMENTED** [review]
Let me know if you'd like to hop on zoom again to discuss any of my comments.  

Thanks!

File: client/src/index.tsx [hunk 16]
```diff
@@ -0,0 +1,39 @@
+import ReactDOM from 'react-dom/client';
+import './index.css';
+import {mutators} from './model/mutators';
+import {Replicache} from 'replicache';
+import {UndoManager} from '@rocicorp/undo';
+import App from './app';
+
+// TODO: see `id.tsx`
```
[the remark this answers is not in the document]
Comment (tantaman) [reply client/src/index.tsx:8]: ah, was for me to review `id.tsx` from the original Repliear to understand how to setup Replicache in the root element. This is done now.

File: client/index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (tantaman) [reply client/index.html]: Fixed here: a0660d5630943af7452ce793a217ba0ae39235f6

File: client/src/util/useLock.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import {useEffect, useState} from 'react';
+import {acquire} from './sync-lock';
+
+/**
+ * Acquires a web lock for the duration of the tab.
+ * Only runs the effect if the lock is held.
+ *
+ * Useful for ensuring only a single table can perform a specific effect.
```
[the remark this answers is not in the document]
Comment (tantaman) [reply client/src/util/useLock.ts]: useLock was removed. We'll be doing this via a lease inside `puller` as discussed over zoom.

File: server/src/push.ts [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    const t1 = Date.now();
+
+    if (error === null) {
+      try {
+        await mutate(executor, mutation);
+      } catch (e) {
+        console.error(
+          `Error executing mutation: ${JSON.stringify(mutation)}: ${e}`,
+        );
+        return {error: String(e)};
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/push.ts:100]: Yep. Also looks like this bug exists in todo-row-versioning: https://github.com/rocicorp/todo-row-versioning/blob/main/server/src/push.ts#L120-L123

File: server/src/pull/pull.ts [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+      const prevCVR = pull.cookie
+        ? await getCVR(executor, pull.cookie.clientGroupID, pull.cookie.order)
+        : undefined;
+      const baseCVR: CVR = prevCVR ?? {
+        clientGroupID,
+        clientVersion: 0,
+        order: 0,
+      };
+
+      // Drop any cvr entries greater than the order we received.
```
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/pull/pull.ts:54]: > What happens if we drop cvr enteries for a cvrversion (i.e. order) and then later getting a pull with that cvrversion (i.e. order)?

We'd compute a new patch from head. The new patch would be recorded with the next cvrversion.

Does that work with respect to Replicache's expectations?

[the remark this answers is not in the document]
Comment (aboodman) [reply server/src/pull/pull.ts:54]: Elsewhere @grgbkr suggested that this implies that when we the pull endpoint receives an old cookie, it should _not_ delete newer entries.

Instead, it should do the diff against client_view_entries that existed as of the received cookie and effectively fast-forward itself to the current state of the database.

This implies that this pull chunk might be bigger than desired but I think that's a good tradeoff.

File: server/src/pull/pull.ts [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+      // To keep greater CVRs around, which may have never been received, means that the next CVR would
+      // indicate that the data was received when it was not.
+      await dropCVREntries(executor, clientGroupID, baseCVR.order);
+
+      const [clientChanges, nextPage] = await Promise.all([
+        searchClients(executor, {
+          clientGroupID,
+          sinceClientVersion: baseCVR.clientVersion,
+        }),
+        readNextPage(executor, clientGroupID, baseCVR.order),
```
[the remark this answers is not in the document]
Comment (aboodman) [reply server/src/pull/pull.ts:69]: Let's match Repliear's current implementation. It's too hard to think about paginating mutations.

[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/pull/pull.ts:69]: It doesn't seem that I can pass an arg to `pull` in order to differentiate the reason why I'm pulling. I.e., pulling in response to a poke or pulling for incremental sync.

File: server/src/pull/pull.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/pull/pull.ts:54]: > Instead, it should do the diff against client_view_entries that existed as of the received cookie and effectively fast-forward itself to the current state of the database.

Which we can do while deleting new entries. On Friday, I thought we landed on:

1. keep the delete
2. remove the <= from `findUnsentItems`
3. " diff against client_view_entries that existed as of the received cookie" which, to me, means: "re-run findUnsentItems with the received cookie"

> Since we dropped the cvr entries we don't know what exactly the client has in its client view to compute a diff against, right?

We still have the entries as of the received cookie. I may be misunderstanding but, to me, if the client sends a cookie then that is telling us what they have. They have all data sent up to and including that cookie value but nothing later. Of course the client _could_ have something later since there _could_ be two tabs fighting but the server can't differentiate between that case or packets being dropped and causing the client to send old cookies.

We don't have to delete new entries but I think that puts us back at needing to make the entries immutable.

File: server/src/pull/cvr.ts (hunk 15 above)
Comment (tantaman) [reply server/src/pull/cvr.ts:93]: Delete does currently miss this case.

File: server/src/pull/pull.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (aboodman) [reply server/src/pull/pull.ts:69]: OK let's talk about this tomorrow.

File: server/src/pull/cvr.ts (hunk 15 above)
Comment (aboodman) [reply server/src/pull/cvr.ts:93]: No it's not expected. We want people to be able to completely delete rows from the primary table. So their row-version will reset.

But that should be fine. `findUnsentItems` is only comparing versions by equality. It doesn't need them to be ordered. Right?

Comment (tantaman) [reply server/src/pull/cvr.ts:93]: Yeah, that's true for `findUnsentItems` and can do the same for delete + re-insert.

I think I see where the disconnect between @grgbkr and I (on https://github.com/rocicorp/repliear/pull/158#discussion_r1414833422) is coming from now that I'm thinking a bit more about delete.

File: server/src/pull/pull.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (tantaman) [reply server/src/pull/pull.ts:54]: I think I see the issue you're getting at (https://github.com/rocicorp/repliear/pull/158#discussion_r1415766295). Let me think on this a bit more 🤔

**Comment by aboodman** [comment]
@tantaman - remind me why we have both https://github.com/rocicorp/repliear/blob/3bb93995c0cc2e06dbc0e1e6e044d6c7c0464b14/server/src/schema.ts#L26 and https://github.com/rocicorp/repliear/blob/3bb93995c0cc2e06dbc0e1e6e044d6c7c0464b14/server/src/schema.ts#L75? It seems like they move together. It seems like we only need the one on `client_view`.

`client_view` has a 1:1 relationship with `replicache_client_group` right? The only real reason for `client_view` right now at all is because we need to track the `client_version` that was last sent to this group. If we can remove that, then this whole table can go away.

But as long as we do have the table, then it seems like we can track the latest version for the `client_view` on its row, rather than on the somewhat unrelated `replicache_client_group`.

**Comment by tantaman** [comment]
> @tantaman - remind me why we have both

Basically it was from my ignorance when starting the conversion. `replicache_client_group` existed in `todo-row-versioning` and I wanted to minimize the moving pieces until everything was working.

I think we can consolidate them now. The rest of the updates we talked about are about done. Writing some tests and tracking down one last bug.

**Comment by tantaman** [comment]
this https://github.com/rocicorp/repliear/pull/158/commits/cfefc45b43ca6a2c4050c967ced9f0748f8fe6e9 has everything we talked about + tests

- fast forwards when receiving an old cookie
- handles deletes and re-insertions of previously deleted rows
- pulls all data that may have been updated or deleted since the last pull in order to preserve consistency semantics for mutations

LMK if you'd rather this be a new PR stack on top of the old PR. That is my preferred workflow (small, stacked, atomic PRs that are easily reviewed) although github doesn't seem to support it well.

I did see some weird behavior with `experimentalWatch` and 4 tabs syncing at once. You can read more in the commit message here: https://github.com/rocicorp/repliear/pull/158/commits/a3a11ac7809bae35b83966186083812dee1d6100

File: notes.md [hunk 21]
```diff
@@ -1,8 +1,6 @@
+# TODO:
 
+- multiple selections for filter infinite loop problem
```
Comment (aboodman) [inline notes.md:3]: can you expand on these, i'm not sure what they mean.

**Comment by aboodman** [comment]
Thanks @tantaman - I reviewed this and overall it looks very great. I'm so stoked to get this landed.

There are a bunch of tiny comments I have though and I think it would make more sense for me to take it from here for two reasons - (a) it's going to be more efficient for to just make all the tiny changes, (b) it's hard to review a change this size really thoroughly in code review. I will feel more confident if I've "had my hands on it" so to speak.

So I think I'll just take the PR over and add the finishing touches. Is that OK with you?

**Comment by aboodman** [comment]
You could move on to integrating materialite.

**Comment by tantaman** [comment]
> So I think I'll just take the PR over and add the finishing touches. Is that OK with you?

Yep, that's fine with me.

Yeah, big PR's are a big pain to deal with. Hopefully we can keep them small for future work.
